src/index.js

```diff
@@ -0,0 +1,157 @@
+const insertLinksIntoHead = require('./lib/insert-links-into-head');
+
+class PreloadPlugin {
+  constructor(options) {
+    this.options = Object.assign({}, defaultOptions, options);
+  }
+
+  addLinks(webpackVersion, compilation, htmlPluginData) {
+    assert(webpackVersion in doesChunkBelongToHTML,
+      `An invalid webpackVersion was supplied. Supported values: ${Object.keys(doesChunkBelongToHTML)}.`);
+
+    const options = this.options;
+
+    // Bail out early if we're configured to exclude this HTML file.
+    if (options.excludeHtmlNames.includes(htmlPluginData.plugin.options.filename)) {
+      return htmlPluginData;
+    }
+
+    const extractedChunks = extractChunks({
+      compilation,
+      optionsInclude: options.include,
+    });
+
+    const htmlChunks = options.include === 'allAssets' ?
+      // Handle all chunks.
+      extractedChunks :
+      // Only handle chunks imported by this HtmlWebpackPlugin.
+      extractedChunks.filter((chunk) => doesChunkBelongToHTML[webpackVersion]({
+        chunk,
+        compilation,
+        htmlAssetsChunks: Object.values(htmlPluginData.assets.chunks),
+      }));
+
+    // Flatten the list of files.
+    const allFiles = htmlChunks.reduce((accumulated, chunk) => {
+      return accumulated.concat(chunk.files);
+    }, []);
+    const uniqueFiles = new Set(allFiles);
+    const filteredFiles = [...uniqueFiles].filter(file => {
+      return (
+        !this.options.fileWhitelist ||
+        this.options.fileWhitelist.some(regex => regex.test(file))
+      );
+    }).filter(file => {
+      return (
+        !this.options.fileBlacklist ||
+        this.options.fileBlacklist.every(regex => !regex.test(file))
+      );
+    });
+    // Sort to ensure the output is predictable.
+    const sortedFilteredFiles = filteredFiles.sort();
+
+    const links = [];
+    const publicPath = compilation.outputOptions.publicPath || '';
+    for (const file of sortedFilteredFiles) {
+      const href = `${publicPath}${file}`;
+
+      const attributes = {
+        href,
+        rel: options.rel,
+      };
+
+      // If we're preloading this resource (as opposed to prefetching),
+      // then we need to set the 'as' attribute correctly.
+      if (options.rel === 'preload') {
+        attributes.as = determineAsValue({
+          href,
+          optionsAs: options.as,
+        });
+
+        // On the off chance that we have a cross-origin 'href' attribute,
+        // set crossOrigin on the <link> to trigger CORS mode. Non-CORS
+        // fonts can't be used.
+        if (attributes.as === 'font') {
+          attributes.crossorigin = '';
+        }
+      }
+
+      const linkElementString = createHTMLElementString({
+        attributes,
+        elementName: 'link',
+      });
+      links.push(linkElementString);
+    }
+
+    htmlPluginData.html = insertLinksIntoHead({
+      links,
+      html: htmlPluginData.html,
+    });
+
+    return htmlPluginData;
+  }
+
+  apply(compiler) {
```

Comment:
You might find this nicer than having to duplicate the hooks:
https://github.com/GoogleChromeLabs/critters/blob/c53df3d6b1f1b63b6374d79b0fa565c48e60a85f/src/util.js#L1

Reply:
That's a good pattern in general.

We ran into issues with folks who listed this plugin prior to the `HtmlWebpackPlugin`, or just didn't list `HtmlWebpackPlugin` at all, and in v4 we can detect that at https://github.com/GoogleChromeLabs/preload-webpack-plugin/pull/72/files/b1f8ec65d655dfa064fad73765c6c552dc089a38#diff-1fdf421c05c1140f6d71444ea2b27638R135 and show a meaningful error message. I think it's worthwhile keeping the tapping logic explicit in this case since we're doing that detection.